### PR TITLE
feat: Add SMTP Service for email delivery ingredient

### PR DESCRIPTION
**Pull request by KShivendu**
## Summary of change

Add SMTP Service for email delivery ingredient and emailpassword + emailverification recipes.

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens_python/constants.py`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `setup.py`
    -   In `supertokens_python/constants.py`
-   [ ] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `supertokens_python/utils.py` file to include that in the `FRAMEWORKS` variable
-   [ ] If added a new recipe that has a User type with extra info, then be sure to change the User type in supertokens_python/types.py

### CODE REVIEW

File: supertokens_python/ingredients/emaildelivery/service/smtp.py [hunk 1]
```diff
@@ -46,68 +47,111 @@ def __init__(
...
-    subject: str
-    toEmail: str
+class GetContentResult:
+    def __init__(self, body: str, subject: str, to_email: str) -> None:
+        self.body = body
+        self.subject = subject
+        self.to_email = to_email
 
 
 class TypeInputSendRawEmailFrom:
```
Comment (bhumilsarvaiya) [inline supertokens_python/ingredients/emaildelivery/service/smtp.py:57]: Do we need this class? Can't we use `SMTPServiceConfigFrom` instead?

File: supertokens_python/ingredients/emaildelivery/service/smtp.py [hunk 2]
```diff
@@ -13,30 +13,31 @@
...
 
 
 class SMTPServiceConfig:
     def __init__(
-        self,
-        host: str,
-        email_from: SMTPServiceConfigFrom,
-        port: int,
-        secure: Union[bool, None] = None,
+        self, host: str, email_from: SMTPServiceConfigFrom,
```
Comment (bhumilsarvaiya) [inline supertokens_python/ingredients/emaildelivery/service/smtp.py:39]: change `email_from` to `from`

File: supertokens_python/ingredients/emaildelivery/service/smtp.py [hunk 3]
```diff
@@ -46,68 +47,111 @@ def __init__(
...
         self.name = name
         self.email = email
 
 
-class TypeInputSendRawEmail(GetContentResult):
-    user_context: Dict[str, Any]
-    config_from: TypeInputSendRawEmailFrom
-
-
 class TypeGetContentInput(Generic[_T]):
```
Comment (bhumilsarvaiya) [inline supertokens_python/ingredients/emaildelivery/service/smtp.py:63]: remove this if not required

File: supertokens_python/ingredients/emaildelivery/service/smtp.py [hunk 4]
```diff
@@ -46,68 +47,111 @@ def __init__(
...
-    return Transporter()
+            smtp.sendmail(config_from.email, get_content_result.to_email, get_content_result.body)
+            smtp.quit()
+        except Exception as _:
+            pass
 
 
 TypeGetDefaultEmailServiceImpl = Callable[[Transporter, TypeInputSendRawEmailFrom], ServiceInterface[_T]]
 
+class SMTPEmailDeliveryImplementation(EmailDeliveryInterface[_T]):
```
Comment (bhumilsarvaiya) [inline supertokens_python/ingredients/emaildelivery/service/smtp.py:110]: remove this if not used

File: supertokens_python/ingredients/emaildelivery/service/smtp.py (hunk 2 above)
Comment (KShivendu) [reply supertokens_python/ingredients/emaildelivery/service/smtp.py:39]: We can't. `from` is a reserved keyword :')

File: supertokens_python/ingredients/emaildelivery/service/smtp.py (hunk 4 above)
Comment (KShivendu) [reply supertokens_python/ingredients/emaildelivery/service/smtp.py:110]: Done. Thanks! 

How do you identify unused code so easily?

Comment (KShivendu) [reply supertokens_python/ingredients/emaildelivery/service/smtp.py:110]: Thanks!

File: supertokens_python/ingredients/emaildelivery/service/smtp.py (hunk 3 above)
Comment (KShivendu) [reply supertokens_python/ingredients/emaildelivery/service/smtp.py:63]: Thanks!

File: supertokens_python/ingredients/emaildelivery/service/smtp.py [hunk 5]
```diff
@@ -46,68 +47,111 @@ def __init__(
...
 class TypeGetContentInput(Generic[_T]):
     user_context: Dict[str, Any]
 
 
 class ServiceInterface(ABC, Generic[_T]):
     @abstractmethod
-    def send_raw_email(self, input: TypeInputSendRawEmail):
+    async def send_raw_email(self,
+                             get_content_result: GetContentResult,
+                             config_from: TypeInputSendRawEmailFrom,
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/ingredients/emaildelivery/service/smtp.py:71]: We can't. from is a reserved keyword :')

File: supertokens_python/ingredients/emaildelivery/service/smtp.py [hunk 6]
```diff
@@ -46,68 +45,69 @@ def __init__(
...
 
-def createTransport(_) -> Transporter:
-    return Transporter()
 
-
-TypeGetDefaultEmailServiceImpl = Callable[[Transporter, TypeInputSendRawEmailFrom], ServiceInterface[_T]]
+TypeGetDefaultEmailServiceImpl = Callable[[Transporter, SMTPServiceConfigFrom], ServiceInterface[_T]]
 
 
 def getEmailServiceImplementation(
```
Comment (bhumilsarvaiya) [inline supertokens_python/ingredients/emaildelivery/service/smtp.py:99]: you can remove this

File: supertokens_python/ingredients/emaildelivery/service/smtp.py (hunk 2 above)
Comment (bhumilsarvaiya) [reply supertokens_python/ingredients/emaildelivery/service/smtp.py:37]: ohh lol, yeah

File: supertokens_python/recipe/emailpassword/api/implementation.py [hunk 7]
```diff
@@ -61,10 +63,11 @@ async def generate_password_reset_token_post(self, form_fields: List[FormField],
         password_reset_link = await api_options.config.reset_password_using_token_feature.get_reset_password_url(
             user, user_context) + '?token=' + token + '&rid=' + api_options.recipe_id
 
-        try:
-            await api_options.config.reset_password_using_token_feature.create_and_send_custom_email(user, password_reset_link, user_context)
-        except Exception:
-            pass
+        inp = TypeEmailPasswordPasswordResetEmailDeliveryInput(
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/api/implementation.py:66]: call this send_email_input

File: supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
+# Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py:1]: create a folder called service_implementation and move implementation.py & email_verification_implementation.py  to that folder. implementation.py will become __init__.py in the new folder

File: supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+from supertokens_python.recipe.emailpassword.emaildelivery.service.smtp.password_reset_implementation import \
+    getPasswordResetEmailContent
+from supertokens_python.recipe.emailpassword.types import \
+    TypeEmailPasswordEmailDeliveryInput
+from supertokens_python.recipe.emailverification.emaildelivery.service.smtp import \
+    getServiceImplementation as \
+    getEmailVerificationEmailDeliveryServiceImplementation
+from supertokens_python.recipe.emailverification.interfaces import \
+    TypeEmailVerificationEmailDeliveryInput
+
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py:28]:
```python
from .email_verification_implementation import DefaultServiceImplementation as DerivedEVServiceImplementation
```

File: supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+class DefaultServiceImplementation(ServiceInterface[TypeEmailPasswordEmailDeliveryInput]):
+    def __init__(self, transporter: Transporter, emailVerificationSeriveImpl: ServiceInterface[TypeEmailVerificationEmailDeliveryInput]) -> None:
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py:31]: emailVerificationSeriveImpl -> email_verification_service_implementation

File: supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        self.transporter = transporter
+        self.emailVerificationSeriveImpl = emailVerificationSeriveImpl
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py:33]: self.emailVerificationSeriveImpl -> self.email_verification_service_implementation

File: supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    async def send_raw_email(self, get_content_result: GetContentResult, config_from: SMTPServiceConfigFrom, user_context: Dict[str, Any]) -> None:
+        self.transporter.send_email(config_from, get_content_result, user_context)
+
+    def get_content(self, email_input: TypeEmailPasswordEmailDeliveryInput, user_context: Dict[str, Any]) -> GetContentResult:
+        if isinstance(email_input, TypeEmailVerificationEmailDeliveryInput):
+            return self.emailVerificationSeriveImpl.get_content(email_input, user_context)  # FIXME: Node SDK has DerivedEV
+        return getPasswordResetEmailContent(email_input)
+
+
+def getServiceImplementation(transporter: Transporter) -> ServiceInterface[TypeEmailPasswordEmailDeliveryInput]:
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py:44]: you can remove this function entirely. Just use DefaultServiceImplementation directly wherever required

File: supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/password_reset_implementation.py [hunk 13]
```diff
@@ -0,0 +1,364 @@
+from supertokens_python.ingredients.emaildelivery.service.smtp import \
+    GetContentResult
+from supertokens_python.recipe.emailpassword.types import \
+    TypeEmailPasswordPasswordResetEmailDeliveryInput
+from supertokens_python.supertokens import Supertokens
+
+
+def getPasswordResetEmailContent(input: TypeEmailPasswordPasswordResetEmailDeliveryInput) -> GetContentResult:
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/password_reset_implementation.py:8]: getPasswordResetEmailContent -> get_password_reset_email_content

File: supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/password_reset_implementation.py [hunk 14]
```diff
@@ -0,0 +1,364 @@
...
+    supertokens = Supertokens.get_instance()
+    app_name = supertokens.app_info.app_name
+    body = getPasswordResetEmailHTML(app_name, input.user.email, input.password_reset_link)
+    content_result = GetContentResult(body, "Email verification instructions", input.user.email)
+    return content_result
+
+
+def getPasswordResetEmailHTML(appName: str, email: str, resetLink: str):
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/password_reset_implementation.py:16]: getPasswordResetEmailHTML -> get_password_reset_email_html

File: supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+from supertokens_python.recipe.emailpassword.types import \
+    TypeEmailPasswordEmailDeliveryInput
+from supertokens_python.recipe.emailverification.emaildelivery.service.smtp import \
+    getServiceImplementation as \
+    getEmailVerificationEmailDeliveryServiceImplementation
+from supertokens_python.recipe.emailverification.interfaces import \
+    TypeEmailVerificationEmailDeliveryInput
+
+
+class DefaultServiceImplementation(ServiceInterface[TypeEmailPasswordEmailDeliveryInput]):
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py:30]: DefaultServiceImplementation -> ServiceImplementation

File: supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/email_verification_implementation.py [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+from typing import Any, Dict
+
+from supertokens_python.ingredients.emaildelivery.service.smtp import (
+    GetContentResult, ServiceInterface, SMTPServiceConfigFrom)
+from supertokens_python.recipe.emailpassword.types import (
+    TypeEmailPasswordEmailDeliveryInput,
+    TypeEmailVerificationEmailDeliveryInput)
+
+
+class DefaultServiceImplementation(ServiceInterface[TypeEmailVerificationEmailDeliveryInput]):
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/email_verification_implementation.py:24]: DefaultServiceImplementation -> ServiceImplementation

File: supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/email_verification_implementation.py [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+    def __init__(self, emailPasswordServiceImpl: ServiceInterface[TypeEmailPasswordEmailDeliveryInput]) -> None:
+        self.emailVerificationSeriveImpl = emailPasswordServiceImpl
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/email_verification_implementation.py:26]: self.emailVerificationServiceImpl -> self.email_verification_service_implementation

File: supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/email_verification_implementation.py [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+
+    async def send_raw_email(self, get_content_result: GetContentResult,
+                             config_from: SMTPServiceConfigFrom, user_context: Dict[str, Any]) -> None:
+        return await self.emailVerificationSeriveImpl.send_raw_email(get_content_result, config_from, user_context)
+
+    def get_content(self, email_input: TypeEmailVerificationEmailDeliveryInput, user_context: Dict[str, Any]) -> GetContentResult:
+        return self.emailVerificationSeriveImpl.get_content(email_input, user_context)
+
+
+def getServiceInterface(emailPasswordServiceImplementation: ServiceInterface[TypeEmailPasswordEmailDeliveryInput]) -> ServiceInterface[TypeEmailVerificationEmailDeliveryInput]:
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/email_verification_implementation.py:36]: you can remove this function entirely. Just use DefaultServiceImplementation directly wherever required

File: supertokens_python/recipe/emailverification/emaildelivery/service/smtp/implementation.py [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+
+from supertokens_python.ingredients.emaildelivery.service.smtp import (
+    GetContentResult, ServiceInterface, SMTPServiceConfigFrom, Transporter)
+from supertokens_python.recipe.emailverification.emaildelivery.service.smtp.email_verify import \
+    getEmailVerifyEmailContent
+from supertokens_python.recipe.emailverification.interfaces import \
+    TypeEmailVerificationEmailDeliveryInput
+
+
+class DefaultServiceImplementation(ServiceInterface[TypeEmailVerificationEmailDeliveryInput]):
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailverification/emaildelivery/service/smtp/implementation.py:13]: DefaultServiceImplementation -> ServiceImplementation

File: supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+from typing import Any, Dict
+
+from supertokens_python.ingredients.emaildelivery.service.smtp import (
+    GetContentResult, ServiceInterface, SMTPServiceConfigFrom, Transporter)
+from supertokens_python.recipe.emailpassword.emaildelivery.service.smtp.password_reset_implementation import \
+    getPasswordResetEmailContent
+from supertokens_python.recipe.emailpassword.types import \
+    TypeEmailPasswordEmailDeliveryInput
+from supertokens_python.recipe.emailverification.emaildelivery.service.smtp import \
+    getServiceImplementation as \
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py:24]: from supertokens_python.recipe.emailverification.emaildelivery.service.smtp import ServiceImplementation as EVServiceImplementation

File: supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py (hunk 10 above)
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py:31]: def __init__(self, transporter: Transporter) -> None:

File: supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py (hunk 11 above)
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py:33]:
```python
email_verification_service_implementation = EVServiceImplementation(transporter)
```

File: supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+    def __init__(self, transporter: Transporter, emailVerificationSeriveImpl: ServiceInterface[TypeEmailVerificationEmailDeliveryInput]) -> None:
+        self.transporter = transporter
+        self.emailVerificationSeriveImpl = emailVerificationSeriveImpl
+
+    async def send_raw_email(self, get_content_result: GetContentResult, config_from: SMTPServiceConfigFrom, user_context: Dict[str, Any]) -> None:
+        self.transporter.send_email(config_from, get_content_result, user_context)
+
+    def get_content(self, email_input: TypeEmailPasswordEmailDeliveryInput, user_context: Dict[str, Any]) -> GetContentResult:
+        if isinstance(email_input, TypeEmailVerificationEmailDeliveryInput):
+            return self.emailVerificationSeriveImpl.get_content(email_input, user_context)  # FIXME: Node SDK has DerivedEV
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/implementation.py:40]: self.emailVerificationSeriveImpl.get_content -> self.ev_get_content

File: supertokens_python/ingredients/emaildelivery/service/smtp.py [hunk 22]
```diff
@@ -46,68 +45,69 @@ def __init__(
...
+    async def send_raw_email(self,
+                             get_content_result: GetContentResult,
+                             config_from: SMTPServiceConfigFrom,
+                             user_context: Dict[str, Any]
+                             ) -> None:
         pass
 
     @abstractmethod
-    def get_content(self, input: TypeGetContentInput[_T]):
+    def get_content(self, email_input: _T, user_context: Dict[str, Any]) -> GetContentResult:
```
Comment (bhumilsarvaiya) [inline supertokens_python/ingredients/emaildelivery/service/smtp.py:65]: also make this a async function

File: supertokens_python/recipe/emailverification/emaildelivery/service/smtp/email_verify.py [hunk 23]
```diff
@@ -0,0 +1,367 @@
+
+
+from supertokens_python.ingredients.emaildelivery.service.smtp import \
+    GetContentResult
+from supertokens_python.recipe.emailverification.interfaces import \
+    TypeEmailVerificationEmailDeliveryInput
+from supertokens_python.supertokens import Supertokens
+
+
+def getEmailVerifyEmailContent(input: TypeEmailVerificationEmailDeliveryInput) -> GetContentResult:
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailverification/emaildelivery/service/smtp/email_verify.py:10]: getEmailVerifyEmailContent -> get_email_verify_email_content

File: supertokens_python/recipe/emailverification/emaildelivery/service/smtp/email_verify.py [hunk 24]
```diff
@@ -0,0 +1,367 @@
...
+    supertokens = Supertokens.get_instance()
+    app_name = supertokens.app_info.app_name
+    body = getEmailVerifyEmailHTML(app_name, input.user.email, input.email_verify_link)
+    content_result = GetContentResult(body, "Email verification instructions", input.user.email)
+    return content_result
+
+
+def getEmailVerifyEmailHTML(app_name: str, email: str, verification_link: str):
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailverification/emaildelivery/service/smtp/email_verify.py:18]: getEmailVerifyEmailHTML -> get_email_verify_email_html

File: supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/__init__.py [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+
+
+class SMTPService(EmailDeliveryInterface[TypeEmailPasswordEmailDeliveryInput]):
+    serviceImpl: ServiceInterface[TypeEmailPasswordEmailDeliveryInput]
+
+    def __init__(self, config: EmailDeliverySMTPConfig[TypeEmailPasswordEmailDeliveryInput]) -> None:
+        self.config = config
+        self.transporter = Transporter(config.smtpSettings)
+        oi = getServiceImplementation(self.transporter)
+        self.serviceImpl = oi if config.override is None else config.override(oi)
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/__init__.py:27]: serviceImpl -> service_implementation

File: supertokens_python/ingredients/emaildelivery/service/smtp.py (hunk 4 above)
Comment (bhumilsarvaiya) [reply supertokens_python/ingredients/emaildelivery/service/smtp.py]: Haha, I missed removing this particular function from node from the commit you're referring. So I thought you would also have this function in your code 😆

File: supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/__init__.py [hunk 26]
```diff
@@ -0,0 +1,44 @@
+from typing import Any, Dict
```
Comment (KShivendu) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/smtp/__init__.py:1]: Add Copyright

File: supertokens_python/recipe/emailverification/emaildelivery/service/smtp/__init__.py [hunk 27]
```diff
@@ -0,0 +1,29 @@
+from typing import Any, Dict
```
Comment (KShivendu) [inline supertokens_python/recipe/emailverification/emaildelivery/service/smtp/__init__.py:1]: Add copyright

File: supertokens_python/recipe/emailverification/emaildelivery/service/smtp/implementation.py [hunk 28]
```diff
@@ -0,0 +1,21 @@
+
```
Comment (KShivendu) [inline supertokens_python/recipe/emailverification/emaildelivery/service/smtp/implementation.py:1]: Add copyright

File: tests/test_email.py [hunk 29]
```diff
@@ -21,7 +21,13 @@
 from pytest import fixture, mark
```
Comment (KShivendu) [inline tests/test_email.py:21]: Add copyright

**Pull request merged**